### PR TITLE
Configurable authorization level (iOS)

### LINKED ISSUE #25: Missing Purpose String in Info.plist File for NSLocationAlwaysUsageDescription
## Platforms
iOS

## Versions
- react-native: 0.58.4

## Description

> Missing Purpose String in Info.plist File - Your app's code references one or more APIs that access sensitive user data. The app's Info.plist file should contain a `NSLocationAlwaysUsageDescription` key with a user-facing purpose string explaining clearly and completely why your app needs the data. 

We get the above warning email from Apple store. I use `LocationWhenInUseUsageDescription` but not  `NSLocationAlwaysAndWhenInUseUsageDescription` and `NSLocationAlwaysUsageDescription`. In other terms we want location data only when app is in the foreground not in background.

Things i tried
- Added strings for both `NSLocationAlwaysAndWhenInUseUsageDescription` and `NSLocationAlwaysUsageDescription`. But this library uses the **always permission** when the strings are present, but we don't want those permission.

What is the best way to handle this situation?

### PR DESCRIPTION
**Pull request by tyronet-sportsbet**
# Overview
As per comments in the [react native issue 20879 thread](https://github.com/facebook/react-native/issues/20879), some organizations just want "when in use" permissions. However, because react native (and this library!) reference the `requestAlwaysAuthorization` selector, Apple requires us to include the `NSLocationAlwaysUsageDescription` in our Info.plist. We're happy to provide this key in our `Info.plist`, but `react-native-geolocation` should NOT always assume that means we _want_ that permission, with no way to override it!

There are workarounds in that thread that involve applying patches that rewrite react native code to not reference that symbol. This is not a workable long term solution, in our specific case we have additional closed-source third party dependencies and some of them may reference this `requestAlwaysAuthorization` selector as well.

A solution that solves for all cases is to have the ability for the application to override the authorization level requested from CoreLocation. This PR implements that. I have added an additional configuration option to the `setRNConfiguration` `GeoConfiguration` object: `authorizationLevel`, which when provided will allow you to override which location services authorization level your application requests from the user. Two possible values, `always` and `whenInUse`.

# Test Plan
A unit test here would need to be in the Objective-C side and I couldn't find any so in the absence of that, I've manually tested it: 

* Without providing the `authorizationLevel` key -> maintains old behaviour: it does.
* Setting `authorizationLevel` to `whenInUse` -> requests "when in use" permissions: it does.
* Setting `authorizationLevel` to `always` -> requests "always" permissions: it does.

--- 

Fixes #25

### CODE REVIEW

File: js/index.js [hunk 1]
```diff
@@ -18,6 +18,7 @@ let updatesEnabled = false;
 
 type GeoConfiguration = {
   skipPermissionRequests: boolean,
+  authorizationLevel: 'always' | 'whenInUse',
```
Comment (matt-oakes) [inline js/index.js]: Can we add an `'auto'` value just incase someone wants to set the default behaviour explicitly in a type safe way.

```suggestion
  authorizationLevel: 'always' | 'whenInUse' | 'auto',
```

File: typings/index.d.ts [hunk 2]
```diff
@@ -9,6 +9,7 @@
 
 export type GeolocationConfiguration = {
   skipPermissionRequests: boolean;
+  authorizationLevel: "always" | "whenInUse"
```
Comment (matt-oakes) [inline typings/index.d.ts]: Same as with the flow types

```suggestion
  authorizationLevel: "always" | "whenInUse" | "auto"
```

**Comment by matt-oakes** [comment]
The CircleCI errors look to be the usual Android flakiness. They should pass when you push the changes 👍

**Comment by tyronet-sportsbet** [comment]
Thanks for the very quick response @matt-oakes ! I've made the requested changes, and as you predicted, the checks have passed 😄

**Comment by matt-oakes** [comment]
Thanks, this looks great! Merging and releasing now :shipit:

**Pull request merged**